### ISSUE TITLE
[BUG] rand_dataframe() is ~7x slower than np.random.rand()

### ISSUE THREAD

**Issue by tpn** [rapidsai/cudf#13128]
**Describe the bug**
`rand_dataframe` is really slow.

```
from cudf.testing.dataset_generator import rand_dataframe
import numpy as np
import pandas as pd

%%time 
df = rand_dataframe([{'dtype': 'int32', 'null_frequency': 0, 'cardinality': 1000}], num_rows)

%%time
df = cudf.DataFrame({"0": pd.Series(np.random.rand(num_rows).astype('int32'))})
```

I get 1.25 seconds for `rand_dataframe()`, and 1.72 *milliseconds* for the numpy approach.

Also, interestingly, as num_rows increases, the numpy approach scales linearly, as you'd expect, whereas the `rand_dataframe()` approach appears to have a very large initial cost regardless of 100 vs 10,000,000 rows being requested.

This issue is being created per @wence-'s suggestion here: https://github.com/rapidsai/cudf/pull/12912/files#r1164209280.

TL;DR is I added a new test that does some 1300 permutations, each requiring a random data frame -- with the numpy approach the tests run "basically instantly", whereas using `rand_dataframe()`, each test takes ~2 seconds to run, which is prohibitive when you're dealing with 1300 tests.

**Comment by shwina** [comment]
Thanks for reporting. 

Looking at a profile (`%%prun -s time`) much of the overhead comes from the use of threading, which can be disabled using `use_threads=False`:

```
In [12]: %time df = rand_dataframe([{'dtype': 'int32', 'null_frequency': 0, 'cardinality': 1000}], num_rows)
CPU times: user 109 ms, sys: 402 ms, total: 511 ms
Wall time: 701 ms

In [13]: %time df = rand_dataframe([{'dtype': 'int32', 'null_frequency': 0, 'cardinality': 1000}], num_rows, use_threads=False)
CPU times: user 61.4 ms, sys: 18.3 ms, total: 79.7 ms
Wall time: 78.8 ms
```

cc: @galipremsagar perhaps do you know when it's appropriate to use threading v/s not here?

**Comment by galipremsagar** [comment]
`rand_dataframe` is written to serve multiple purposes tailored for our fuzz-testing framework. Coming to usage of threading, we usually have 100s of columns to be generated at times, for which using each thread to generate a separate column speeds up the process. Ideally we shouldn't be using it for pytests. For pytests, we have simpler APIs like `cudf.datasets.randomdata`.
